### PR TITLE
Federated model

### PR DESCRIPTION
**Pull request by derikclive**
PR adds the initial code for training a federated transformer based model to predict slots and intents in queries. The evaluation code for the federated model is yet to be added.

### CODE REVIEW

File: Federated Model (seq-to-seq)/train_model.py [hunk 1]
```diff
@@ -0,0 +1,358 @@
...
+
+    model = keras.Model(inputs=[
+        sent_input, slot_input, padding_mask, look_ahead_mask, intent_mask
+    ],
+                        outputs=[slot_output, intent_output])
+
+    return model
+
+
+def preprocess(dataset, arg):
```
Comment (jiteshpunjabi) [inline Federated Model (seq-to-seq)/train_model.py:251]: Function is missing a comment.

File: Federated Model (seq-to-seq)/train_model.py [hunk 2]
```diff
@@ -0,0 +1,358 @@
...
+    intent_real: The ground through for the intents.
+    slot_pred: The slots predicted by the model.
+    intent_pred: The intents predicted by the model.
+    intent_loss_objective: The objective used to compute the intent loss.
+    slot_loss_objective: The objective used to compute the slot loss.
+
+    Returns:
+    The total loss.
+    """
+    loss_object = tf.keras.losses.SparseCategoricalCrossentropy(
```
Comment (jiteshpunjabi) [inline Federated Model (seq-to-seq)/train_model.py]: nit: change to loss_objective

File: Federated Model (seq-to-seq)/train_model.py [hunk 3]
```diff
@@ -0,0 +1,358 @@
...
+        from_logits=True, reduction='none')
+    mask = tf.math.logical_not(tf.math.equal(y_true, 0))
+    slot_loss_ = loss_object(y_true, y_pred)
```
Comment (jiteshpunjabi) [inline Federated Model (seq-to-seq)/train_model.py]: nit: change to slot_losses?

File: Federated Model (seq-to-seq)/train_model.py [hunk 4]
```diff
@@ -0,0 +1,358 @@
...
+    #     arg, arg.valid_data_path, in_vocab, slot_vocab, intent_vocab)
+
+    # test_in_data, test_slot_data, test_intent_data = load_dataset(
+    #     arg, arg.test_data_path, in_vocab, slot_vocab, intent_vocab)
+
+    targets = train_slot_data[:, 1:]
+    padding_masks, look_ahead_masks, intent_masks = create_masks(
+        train_in_data, targets)
+
+    num_clients = 10
```
Comment (jiteshpunjabi) [inline Federated Model (seq-to-seq)/train_model.py]: Let's make this an argument.

File: Federated Model (seq-to-seq)/train_model.py [hunk 5]
```diff
@@ -0,0 +1,358 @@
...
+    ftrain_data = collections.OrderedDict()
+    instance_per_client = len(train_in_data) // num_clients
+
+    for i in range(num_clients):
```
Comment (jiteshpunjabi) [inline Federated Model (seq-to-seq)/train_model.py]: We can submit this as is for now but in a quick follow up change, can you please add some code to persist the split training set per client so that we can do some exploratory analyses on them?

File: Federated Model (seq-to-seq)/train_model.py [hunk 6]
```diff
@@ -0,0 +1,358 @@
...
+    example_dataset = preprocess(raw_example_dataset, arg)
+    print(example_dataset.element_spec)
+
+    ftrain_data = make_federated_data(ftrain_data, np.arange(10), arg)
+
+    iterative_process = tff.learning.build_federated_averaging_process(
+        lambda: create_tff_model(
+            arg, len(in_vocab['vocab']), len(slot_vocab['vocab']),
+            len(intent_vocab['vocab']), example_dataset.element_spec),
+        client_optimizer_fn=lambda: tf.keras.optimizers.SGD(learning_rate=0.02
```
Comment (jiteshpunjabi) [inline Federated Model (seq-to-seq)/train_model.py]: Accept learning rates for both the client and the server as command line arguments.

File: Federated Model (seq-to-seq)/train_model.py [hunk 7]
```diff
@@ -0,0 +1,358 @@
...
+                                                            ),
+        server_optimizer_fn=lambda: tf.keras.optimizers.SGD(learning_rate=1.0))
+
+    state = iterative_process.initialize()
+
+    num_rounds = 10
```
Comment (jiteshpunjabi) [inline Federated Model (seq-to-seq)/train_model.py]: This too should be a command line argument.

**Review by jiteshpunjabi — CHANGES_REQUESTED** [review]
Hi Derik,

Thanks for working on this. I have added some high level comments. Mainly style nits.

Can we split this PR in the following order:
1. Utils+Tests for utils.
2. train_model.py

Thanks!

File: Federated Model (seq-to-seq)/train_model.py (hunk 5 above)
Comment (derikclive) [reply Federated Model (seq-to-seq)/train_model.py]: I'll just seed the numpy random generator, if that's alright?

**Comment by jiteshpunjabi** [comment]
Thanks for all the work on this! Merging.
